pycontrails/core/models.py

```diff
@@ -731,25 +762,271 @@ def interpolate_met(
+
+
+def _prepare_q(  # type: ignore[return-value]
+    mda: MetDataArray, level: npt.NDArray[np.float_], q_method: str, log_applied: bool
+) -> tuple[MetDataArray, npt.NDArray[np.float_]]:
+    """Prepare specific humidity for interpolation with experimental ``q_method``.
+
+    Parameters
+    ----------
+    mda : MetDataArray
+        MetDataArray of specific humidity.
+    level : npt.NDArray[np.float_]
+        Levels to interpolate to, [:math:`hPa`].
+    q_method : str
+        One of ``"log-q-log-p"`` or ``"cubic-spline"``.
+    log_applied : bool
+        Whether a log transform was applied to ``mda``.
+
+    Returns
+    -------
+    mda : MetDataArray
+        MetDataArray of specific humidity transformed for interpolation.
+    level : npt.NDArray[np.float_]
+        Transformed levels for interpolation.
+    """
+    da = mda.data
+
+    if q_method == "log-q-log-p":
+        if not da._in_memory:
+            da.load()
+
+        da = da.assign_coords(level=np.log(da["level"]))
+
+        if not log_applied:
+            # ERA5 specific humidity can have negative values
+            # These will get converted to NaNs
+            # Ignore the xarray warning
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", message="invalid value encountered in log")
+                da = np.log(da)  # type: ignore[assignment]
+
+        mda = MetDataArray(da, copy=False)
+
+        level = np.log(level)
+        return mda, level
+
+    assert not log_applied, "Log transform should not be applied for cubic spline interpolation"
+
+    if q_method == "cubic-spline":
+        if da["level"][0] < 50.0 or da["level"][-1] > 1000.0:
+            raise ValueError("Cubic spline interpolation requires data to span 50-1000 hPa.")
+        ppoly = _load_spline()
+
+        da = da.assign_coords(level=ppoly(da["level"]))
+        mda = MetDataArray(da, copy=False)
+
+        level = ppoly(level)
+
+        return mda, level
+
+    raise_invalid_q_method_error(q_method)
+
+
+def interpolate_gridded_specific_humidity(
+    mda: MetDataArray,
+    vector: GeoVectorDataset,
+    q_method: str,
+    log_applied: bool,
+    **interp_kwargs: Any,
+) -> np.ndarray:
+    """Interpolate specific humidity against ``vector`` with experimental ``q_method``.
+
+    Parameters
+    ----------
+    mda : MetDataArray
+        MetDataArray of specific humidity.
+    vector : GeoVectorDataset
+        Flight or GeoVectorDataset instance
+    q_method : {"linear-q", "cubic-spline", "log-q-log-p"}
+        Experimental method to use for interpolating specific humidity.
+    log_applied : bool
+        Whether or not a log transform was applied to specific humidity.
+    **interp_kwargs : Any,
+        Additional keyword only arguments passed to `intersect_met`.
+
+    Returns
+    -------
+    np.ndarray
+        Interpolated values.
+    """
+    if q_method == "linear-q":
```

Comment:
This may be clearer if you had it separated by method from the start. There may be some shared logic I'm not seeing, but I think you could have your `extract_q` and `prepare_q` separately here depending on method.

In particular I don't like the `log_applied` boolean getting passed, but I understand if you need the flag.

Reply:
See comment above ... I'm going to keep the log_applied flag for now.